src/app/clusters/icd-management-server/icd-management-server.cpp

```diff
@@ -67,43 +72,173 @@ CHIP_ERROR IcdManagementAttributeAccess::Read(const ConcreteReadAttributePath &
+        InteractionModel::Status status = CheckAdmin(commandObj, is_admin);
+        VerifyOrReturnError(InteractionModel::Status::Success == status, status);
+
+        if (is_admin)
+        {
+            // Administrator
+            if (commandData.verificationKey.HasValue())
+            {
+                VerifyOrReturnError(commandData.verificationKey.Value().data_equal(entry.key), InteractionModel::Status::Failure);
+            }
+        }
+        else
+        {
+            // Not admin
+            VerifyOrReturnError(commandData.verificationKey.HasValue(), InteractionModel::Status::Failure);
+            VerifyOrReturnError(commandData.verificationKey.Value().data_equal(entry.key), InteractionModel::Status::Failure);
+        }
+    }
+    else if (CHIP_ERROR_NOT_FOUND == err)
+    {
+        // New entry
+        VerifyOrReturnError(entry.index < table.Limit(), InteractionModel::Status::ResourceExhausted);
+    }
+    else
+    {
+        // Error
+        return InteractionModel::Status::Failure;
+    }
+
+    // Save
+    entry.checkInNodeID    = commandData.checkInNodeID;
+    entry.monitoredSubject = commandData.monitoredSubject;
+    entry.key              = commandData.key;
+    err                    = table.Set(entry.index, entry);
+    VerifyOrReturnError(CHIP_NO_ERROR == err, InteractionModel::Status::Failure);
+
+    // Response
+    IcdManagement::Commands::RegisterClientResponse::Type response;
+    chip::app::Clusters::IcdManagement::Attributes::ICDCounter::Get(0, &response.ICDCounter);
+    commandObj->AddResponse(commandPath, response);
+
+    return InteractionModel::Status::Success;
 }
 
 InteractionModel::Status IcdManagementServer::UnregisterClient(
     chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
     const chip::app::Clusters::IcdManagement::Commands::UnregisterClient::DecodableType & commandData)
 {
-    // TODO: Implementent logic for end device
-    return InteractionModel::Status::UnsupportedCommand;
+    FabricIndex fabric = commandObj->GetAccessingFabricIndex();
+    IcdMonitoringTable table(&chip::Server::GetInstance().GetPersistentStorage(), fabric);
+
+    // Get current entry, if exists
+    IcdMonitoringEntry entry;
+    CHIP_ERROR err = table.Find(commandData.checkInNodeID, entry);
+    VerifyOrReturnError(CHIP_ERROR_NOT_FOUND != err, InteractionModel::Status::NotFound);
+    VerifyOrReturnError(CHIP_NO_ERROR == err, InteractionModel::Status::Failure);
+
+    // Get admin status
+    bool is_admin                   = false;
+    InteractionModel::Status status = CheckAdmin(commandObj, is_admin);
+    VerifyOrReturnError(InteractionModel::Status::Success == status, status);
+
+    // Validations
+    if (is_admin)
+    {
+        // Administrator
+        if (commandData.key.HasValue())
+        {
+            VerifyOrReturnError(commandData.key.Value().data_equal(entry.key), InteractionModel::Status::Failure);
+        }
```

Comment:
same comment as `RegisterClient`

Reply:
The verification key is now ignored if the ISD has administrator permissions.